### DISCUSSION TITLE
async modelTransform

### DISCUSSION THREAD

**Discussion by danyhiol** [vazco/uniforms#1036] [Q&A] [answered]
Using a promise with `modelTransform` does not seems to work. 
I am firing some `GraphQL` queries inside the model transform that requires to await for the query. This causes the form to have empty values on the `ui`. 
I am wondering if there is a way to get `modeltransform` to work that way.

**Comment by radekmie** [comment, marked as answer]
Hi @danyhiol. No, there's no way of making `modelTransform` actually capable of handling asynchronous operations. What exactly are you trying to achieve? Maybe there's a better solution.

**Reply by danyhiol** [reply]
I am trying to set a `field value` base on `another field`. The problem is that the value of the fiel itself is the `id` of the `object`. So I've to query the object before setting the value of the field.
Model in:
```{ payment: 'some_id', amount: number, ... };```
Model Out: 
```{ payment: 'some_id', amount: 200, ... };``` where `200` is comming from the payment object that I need to query from the server.
Since I don't know what value `payment` may have by the time the model changed, i decided to use `modelTransform` for that. I've tried to make it async, but it doen't work.

**Reply by radekmie** [reply]
Is this value supposed to be editable for the user?
* If not, then you can download it in the background and display it separately.
* If yes, then what should happen if the user will change the value before it's downloaded? Maybe downloading it in the background and updating the `model` from the outside will be enough?

**Reply by danyhiol** [reply]
Payment is a select field meaning that the value can be changed by the user. For now I am downloading the whole payment db just for filtering. I was wondering if there is another way to handle this. 
Your suggestion looks okay. But how to implement

**Reply by radekmie** [reply]
By _value_ I meant the `amount`, not `payment` - sorry for not being clear here. Whenever a user changes the `payment` field, you may download the actual `amount` value (maybe an entire object, if there's no other way), store it in an external state, and use it during rendering.

The other question that you've just raised - how to populate selection options from the API - was already answered multiple times, depending on the context. In short, create a custom field and handle the options (suggestions) without thinking about uniforms.

**Reply by danyhiol** [reply]
Hi @radekmie, thanks for the reply. 
The `amount` is coming from the payment object. The user cannot change it by himself.
Maybe I am the one that was not clear here. I use generics for `forms`. I don't use a single form for each operation. That would be overkill for the app. So the system has to be designed in a way that a component just calls the form with some props. 
Now fetching the payment depends on its actual value and cannot be done without `modelTransform` being `async` as I am using `ApolloJS(GraphQL)` for that. Changing the model requires `modelTransform`as far as I know. 
Using a parallel operation and rendering it afterwards, also requires a custom or specific form. Since I have only one for my app, that may be complicated to achieve. The best would have been to have `modelTransform`working like `onValidate` meaning async. Right now I am downloading all the payments and filtering mainly because I have to wait for the query before rendering the form with the right amount.

For populating the `SelectFiel` I already have a solution for that. What I was trying to say is that: for me to get the `amount` for the payment, I had to get all payments, and then filter the one selected by the user. 

Maybe my question is more on my app architecture than uniforms itself.

**Reply by radekmie** [reply]
Maybe that's something you could refactor. What is the `amount` needed for? If it depends on the `payment`, it could be validated solely on the backend.

**Reply by danyhiol** [reply]
Yes, it depends on the `payment` it is just to show the user and saved in the DB. It has no real functionality.

**Reply by radekmie** [reply]
In this case, you can render it as a client-only effect e.g., using a field that receives `payment` and fetches the `amount` from the background) and on the server fetch it directly, not to let the user fake it.
